### PR TITLE
Feature/arealight thresolding and error d3 d 2283

### PR DESCRIPTION
**Pull request by Kuranes**
- Ponders luminance by solid angle
- reduce light area size after zone light merge to select area light near analytic directional point
- operate on smaller env map
- reduce a bit float error by using normalized to 0-0.5 values

### CODE REVIEW

**Comment by Kuranes** [comment]
TODO: Latest float error fixing miss "pondering"

File: extractLightsMerge.cpp [hunk 1]
```diff
@@ -3,8 +3,15 @@
 /**
  * convert Env map Regions to Lights
  */
-void createLightsFromRegions(const SatRegionVector& regions, LightVector& lights, const float *rgba, const int width, const int height, const int nc)
+void createLightsFromRegions(const SatRegionVector& regions, LightVector& lights, const float *rgba, const int width, const int height, const int nc, const SummedAreaTable &lum_sat)
```
Comment (cedricpinson) [inline extractLightsMerge.cpp]: lum_sat -> coding style -> lumSat

File: process_environment.py [hunk 2]
```diff
@@ -471,7 +471,16 @@ def extract_light(self):
         print output_log
 
     def extract_lights(self):
-        cmd = "{} {}".format(extractLights_cmd, self.panorama_highres)
+
+        # compute it one time for panorama
+        img_size_x = 1024
```
Comment (cedricpinson) [inline process_environment.py:476]: do you think we could do it in half resolution ? did you test it.

Comment (Kuranes) [inline process_environment.py:476]: half resolution as in 512x1024 ?
yes, tried but on our biggest area lights, it gave different light direction (not much but still)
( as the smaller the image, the biggest the "cone" of the light direction becomes. )

File: extractLightsMerge.cpp (hunk 1 above)
Comment (Kuranes) [inline extractLightsMerge.cpp]: ok

**Comment by Kuranes** [comment]
TODO: 
- ~~add functionnal test cubemap/envlightextraction~~

**Comment by Kuranes** [comment]
Added functionnal tests:

```
./specularCubemapTest.sh
./extractLightsTest.sh
```

File: process_environment.py (hunk 2 above)
Comment (cedricpinson) [inline process_environment.py:476]: ok

**Comment by Kuranes** [comment]
TODO:
- output debug image on --debug flag on extractLight
- run tests on docker/docker build

**Comment by Kuranes** [comment]
rebased, fixed default param in process_envirronment.py

**Comment by Kuranes** [comment]
Added lum_ratio as in (luminance of that light  / total env map luminance)
So that we can scale our light when we apply it. 
As in 
a 0.20 light cannot remove more that 20% of total Env lighting computed

**Comment by Kuranes** [comment]
- rebased latest develop
- removed NdotL from light substraction equation
- uses latest envtools with lum_ratio

**Comment by cedricpinson** [comment]
Tests script does not work if the build directory is not the root directory of the project.
like if you build into a directory build you need to do manual stuff to run test.
Could you make a target like make test and then everything is tested whatever the build directory is  please

**Comment by Kuranes** [comment]
- added "make check"
  (it's not "make test" as "test" is cmake reserved)

**Comment by cedricpinson** [comment]
it does not work

trigrou@bim:~/dev/envtools # mkdir b2
trigrou@bim:~/dev/envtools # cd b2
trigrou@bim:~/dev/envtools/b2 # cmake ../
-- The C compiler identification is AppleClang 7.3.0.7030031
-- The CXX compiler identification is AppleClang 7.3.0.7030031
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Try OpenMP C flag = [-fopenmp=libomp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [ ]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [/openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-Qopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-xopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [+Oopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-qsmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP C flag = [-mp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-fopenmp=libomp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [ ]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [/openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-Qopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-xopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [+Oopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-qsmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Try OpenMP CXX flag = [-mp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Failed
-- Could NOT find OpenMP (missing:  OpenMP_C_FLAGS OpenMP_CXX_FLAGS) 
-- Boost version: 1.61.0
-- Found the following Boost libraries:
--   system
-- Found TBB: /usr/local/include (found version "4.3")  
-- Found ZLIB: /usr/lib/libz.dylib (found version "1.2.5") 
-- Found PNG: /usr/local/lib/libpng.dylib (found version "1.6.25") 
-- Found TIFF: /usr/local/lib/libtiff.dylib (found version "4.0.6") 
-- Found JPEG: /usr/local/lib/libjpeg.dylib  
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/trigrou/dev/envtools/b2
trigrou@bim:~/dev/envtools/b2 # make ; make check
Scanning dependencies of target envBRDF
[  3%] Building CXX object CMakeFiles/envBRDF.dir/envBRDF.cpp.o
[  7%] Linking CXX executable envBRDF
[  7%] Built target envBRDF
Scanning dependencies of target envremap
[ 10%] Building CXX object CMakeFiles/envremap.dir/envremap.cpp.o
[ 14%] Linking CXX executable envremap
[ 14%] Built target envremap
Scanning dependencies of target panoramaPacker
[ 17%] Building CXX object CMakeFiles/panoramaPacker.dir/panoramaPacker.cpp.o
[ 21%] Linking CXX executable panoramaPacker
[ 21%] Built target panoramaPacker
Scanning dependencies of target extractLights
[ 25%] Building CXX object CMakeFiles/extractLights.dir/extractLightsVariance.cpp.o
[ 28%] Building CXX object CMakeFiles/extractLights.dir/SummedAreaTable.cpp.o
[ 32%] Building CXX object CMakeFiles/extractLights.dir/SummedAreaTableRegion.cpp.o
[ 35%] Linking CXX executable extractLights
[ 35%] Built target extractLights
Scanning dependencies of target extractLight
[ 39%] Building CXX object CMakeFiles/extractLight.dir/extractLight.cpp.o
[ 42%] Building CXX object CMakeFiles/extractLight.dir/Cubemap.cpp.o
[ 46%] Linking CXX executable extractLight
[ 46%] Built target extractLight
Scanning dependencies of target samplesGGX
[ 50%] Building CXX object CMakeFiles/samplesGGX.dir/samplesGGX.cpp.o
[ 53%] Building CXX object CMakeFiles/samplesGGX.dir/Cubemap.cpp.o
[ 57%] Linking CXX executable samplesGGX
[ 57%] Built target samplesGGX
Scanning dependencies of target envBackground
[ 60%] Building CXX object CMakeFiles/envBackground.dir/envBackground.cpp.o
[ 64%] Building CXX object CMakeFiles/envBackground.dir/Cubemap.cpp.o
[ 67%] Linking CXX executable envBackground
[ 67%] Built target envBackground
Scanning dependencies of target envIrradiance
[ 71%] Building CXX object CMakeFiles/envIrradiance.dir/envIrradiance.cpp.o
[ 75%] Building CXX object CMakeFiles/envIrradiance.dir/Cubemap.cpp.o
[ 78%] Linking CXX executable envIrradiance
[ 78%] Built target envIrradiance
Scanning dependencies of target envPrefilter
[ 82%] Building CXX object CMakeFiles/envPrefilter.dir/envPrefilter.cpp.o
[ 85%] Building CXX object CMakeFiles/envPrefilter.dir/Cubemap.cpp.o
[ 89%] Linking CXX executable envPrefilter
[ 89%] Built target envPrefilter
Scanning dependencies of target cubemapPacker
[ 92%] Building CXX object CMakeFiles/cubemapPacker.dir/cubemapPacker.cpp.o
[ 96%] Building CXX object CMakeFiles/cubemapPacker.dir/Cubemap.cpp.o
[100%] Linking CXX executable cubemapPacker
[100%] Built target cubemapPacker
Scanning dependencies of target check
./specularCubemapTest.sh: line 7: ./process_environment.py: No such file or directory
oiiotool ERROR: read : File does not exist: "testData/specular_4.tif"
[OK] output diff test OK
Cannot open -d image file
oiiotool ERROR: read : File does not exist: "testData/debug_variance.png"
[ERROR] extractlight output diff changed
Built target check
trigrou@bim:~/dev/envtools/b2 #

**Comment by Kuranes** [comment]
- added missing file to copy steps (install to bin !== copy build dir steps) 
- fix directory copy path wrong level

**Comment by Kuranes** [comment]
- fixed error on diff
- mega tested 2 times on each step (rm -Rf build src git clone  cmake )

**Comment by Kuranes** [comment]
- switched extractLightsTest from bash to python so that we compare json value as float for 0.001 difference in values (pass test on)
- do we keep the debug_variance png in the repository ? (as reference pic ?)

**Comment by Kuranes** [comment]
@groot test finished all good.

Just the question: do we keep the debug_variance png in the repository ? (as reference pic ?)

**Comment by cedricpinson** [comment]
Yes keep it for now

**Pull request merged**